app/src/main/java/com/codesoom/assignment/controllers/TaskController.java

```diff
@@ -0,0 +1,244 @@
+        return task;
+    }
+
+    //@RequestMapping(path = "", method = RequestMethod.POST)
+    @PostMapping
+    public Task create(@RequestBody Task task) {
+        //String body = getBody(exchange);
+
+        //String title = "test";
+        //Task task = toTask(body);
+        //Task task = new Task();
+        task.setId(generate());
+        //task.setTitle(title);
+        tasks.add(task);
+
+        return task;
+    }
+
+    /**
+     * 1. 이해
+     * 우리가 풀어야 될 문제가 무엇인지?
+     *   * List<Task>에서 주어진 Task로 찾아야 한다.
+     *
+     * 우리가 알고있는 사실들은 어떤 것들이 있는지
+     *   * List 타입이다.
+     *   * 주어진 Task에 id가 있다.
+     *
+     * 2. 계획
+     * for(int i; i < List.size; i++) {
+     *     List.get(i).getId() == Task.getId();
+     * }
+     *
+     * List 해시맵으로 변경한다.
+     * 해시맵에서 get한다.
+     *
+     * List를 stream으로 변경한 후 find를 사용한다.
+     *
+     * indexOf를 사용하자
+     /
+
+    /**
+     * 1. 이해
+     * 우리가 풀어야 될 문제가 무엇인지?
+     *   - 할 일의 텍스트를 수정해야 한다.
+     *
+     * 우리가 알고있는 사실들은 어떤 것들이 있는지
+     *   - 저장되어 있는 할 일 목록이 있다. 할 일 목록은 List<Task>다.
+     *   - Spring Annotation을 활용할 수 있다.
+     *   - Task가 이미 정의되어 있다.
+     *   - 요청으로는 Task를 요청한다.
+     *   - 응답으로는 수정된 Task를 응답한다.
+     *   - Task의 setTitle 메서드로 타이틀을 수정할 수 있다.
+     *   - HTTP Method는 PATCH로 요청이 온다.
+     *
+     * 우리가 아직 모르는 것들
+     *   - 사용자가 입력하는 데이터는 어떤 모양인가?
+     *   - 어떤 Task를 수정할지 어떻게 알 수 있을까?
+     *     - Task에 id를 포함해서 전송해야 한다.
+     *
+     * 비슷한 문제를 문제를 풀어본 적이 있는지?
+     *   - 회사에서 데이터를 수정하는 일을 많이 해봤다.
+     *
+     * 우리가 문제를 풀 때 어떤 조건들이 있는지?
+     *   - 수정할 할 일을 찾을 수 없으면 수정할 수 없다.
+     *   - 빈 텍스트로 수정할 수 없다.
+     *
+     * 2. 계획
+     * 사용자의 요청에 대해서 제약(빈 텍스트로 수정할 수 없다, 찾을 수 없으면 수정할 수 없다)을 건다.
+     * 텍스트를 서버로 넘기는 작업
+     * 받은 텍스트를 기존 텍스트를 대체한다.
+     * 대체한 값을 응답한다.
+     *
+     * Task를 받는다.
+     * 우리가 가진 할 일 목록에서, Task를 찾아요.
+     *   * 할 일 목록에서 어떻게 찾는가? => indexOf()찾는다.
+     * 찾은 Task를 수정해요.
+     * 수정된 Task를 응답한다.
+     *
+     * 3. 실행
+     * 4. 회고
+     * 어떻게 하면 더 효과적으로 해결할 수 있었을까?
+     *   찾는 부분을 분리하면 어떨까?
+     *   indexOf를 좀 더 정확하게 파악하고 했으면 좋을 것 같다.
+     *   equals에 대해서 더 정확하게 알고 했다면 더 좋았을 것 같다.
+     * 어떻게 하면 더 효율적으로 해결할 수 있었을까?
+     * 여기서 풀었던 방법으로 다른 곳에도 사용할 수 있을까?
+     *   찾는 부분을 잘 분리하면 삭제할 때도 쓸 수 있지 않을까?
+     *   할 일 목록 중에 하나만 찾아서 반환하는 곳에도 쓸 수 있을 것 같다.
+     */
+    @PatchMapping
+    public Task update(@RequestBody Task task){
+        // * Task를 받는다.
+        // * 우리가 가진 할 일 목록에서, Task를 찾아요
+        //   * 할 일 목록에서 어떻게 찾는가? => indexOf()로 찾는
+        int taskIndex = tasks.indexOf(task);
+        // TODO: 찾을 수 없었을 때 어떻게 할 것인가?
+        // 찾을 수 없는 경우 -1 반환. IndexOutOfBoundsException 오류 발생.
+        // IndexOutOfBoundsException 오류 해결하기 위해 if문 사용.
+        // if문 조건식인 index 값이 -1이 아닌 경우 문장 수행
+        Task foundTask = null;
+        if(taskIndex != -1){
+            tasks.get(taskIndex);
+            // * 찾은 Task를 수정해요.
+            foundTask.setTitle(task.getTitle());
+        }
+        // * 수정된 Task를 응답한다.
+        return foundTask;
```

Comment:
찾을 수 없을 경우 `null`을 반환하여 아무것도 반환하지 않는다 라고 결정하신 것 같아요. 그래서 foundTask를 null로 초기화를 해놓고, 할일을 찾은 경우에만 foundTask의 값을 갱신하여 title을 수정하려고 하신 것 같습니다.

 그런데 지금 null로 초기화를 하고 찾았을 경우 foundTask에 다시 재할당을 하지 않아 Null point exception이 발생할 것 같아요. 따라서 다음과 같이 해야할 것 같습니다.
```suggestion
        Task foundTask = null;
        if(taskIndex != -1){
            foundTask = tasks.get(taskIndex);
            foundTask.setTitle(task.getTitle());
        }

        return foundTask;
```

위의 방법도 좋은 방법이지만, foundTask가 처음에 null로 초기화되고 다시 재할당하는 방법은 변수가 언제 어디서 바뀔지 예측하기 어렵다는 단점이 있습니다. 지금같은 경우는 코드가 7줄 정도밖에 되지 않아서 이해할 수 있지만, 나중에 중간에 다른 복잡한 로직이 껴지면 `변수가 변할 수 있다` 따라서 코드를 읽을 때 기억하며 읽어야 하기 때문에 복잡성을 증가시킵니다. 따라서 찾을 수 없는 경우 빠르게 로직을 종료하여 더 이상 코드를 읽을 필요 없게하고, 변수가 변하지 않는 상수이기 때문에 변하는 것에 대한 두려움 없이 코드를 읽을 수 있도록 할 수 있습니다.

```java
if(taskIndex < 0) {
    return null;
}

Task foundTask = tasks.get(taskIndex);
foundTask.setTitle(task.getTitle());

return foundTask;
```